jax/_src/stages.py

```diff
@@ -179,9 +110,94 @@ def __call__(self, *args, **kwargs):
         if isinstance(arg, core.Tracer):
           raise TypeError(
-              'Cannot apply JAX transformations to a function lowered and '
-              'compiled for a particular signature. Detected argument of '
-              f'Tracer type {type(arg)}.')
+              "Cannot apply JAX transformations to a function lowered and "
+              "compiled for a particular signature. Detected argument of "
+              f"Tracer type {type(arg)}.")
       else:
         raise
     return tree_unflatten(self.out_tree, out_flat)
+
+
+class Lowered:
+  """Lowering of a function specialized to argument types and values.
+
+  A lowering is a computation ready for compilation. This class
+  carries a lowering together with the remaining information needed to
+  later compile and execute it. It also provides a common API for
+  querying properties of lowered computations across JAX's various
+  lowering paths (``jit``, ``pmap``, etc.).
+  """
+  __slots__ = [
+      "in_tree", "in_avals", "out_tree", "donate_argnums", "_lowering",
+      "_no_kwargs"
+  ]
+
+  # The PyTreeDef of the (positional arguments, keyword arguments).
+  #
+  # To get the individual PyTreeDef for the positional an keyword arguments,
+  # use `in_tree.children() which will return you a sequence of 2 PyTreeDef.
+  in_tree: PyTreeDef
+  # The nested input tree of `ShapedArray` abstract values of (args, kwargs).
+  in_avals: Any
+  out_tree: PyTreeDef
+  donate_argnums: Tuple[int]
+  _lowering: Union[dispatch.XlaComputation,
+                   pxla.MeshComputation,
+                   pxla.PmapComputation]
+  _no_kwargs: bool
+
+  def __init__(self,
+               lowering,
+               in_tree: PyTreeDef,
+               in_avals,
+               out_tree: PyTreeDef,
+               donate_argnums: Tuple[int],
+               no_kwargs: bool = False):
+    """Initializer.
+
+    Args:
+      in_tree: The `PyTreeDef` of (args, kwargs).
+      out_tree: The `PyTreeDef` of the outputs.
+      no_kwargs: If `True` the transformation, and the `Compiled` returned from
+        this object will not support keyword arguments (an error will be raised
+        if some are provided).
+    """
+    self._lowering = lowering
+    self.in_tree = in_tree
+    self.in_avals = in_avals
+    self.out_tree = out_tree
+    self.donate_argnums = donate_argnums
+    self._no_kwargs = no_kwargs
+
+  def compile(self) -> Compiled:
+    return Compiled(
+        self._lowering.compile(), self.in_tree, self.in_avals,
+        self.out_tree, self.donate_argnums, self._no_kwargs)
+
+  def compiler_ir(self, dialect: Optional[str] = None):
+    if dialect is None or dialect == "mhlo":
+      return self._lowering.mhlo()
+    elif dialect == "hlo":
+      return self._lowering.hlo()
+    else:
+      raise ValueError(f"Unknown dialect {dialect}")
+
+  # TODO(frostig): remove this in favor of `compiler_ir`
+  def _xla_computation(self):
+    return self._lowering.hlo()
+
+
+class Wrapped(Protocol):
```

Comment:
Maybe something more descriptive and less generic, like `WrappedForStaging` or `StagedCallable`?